File: src/waypalm-site/src/main/java/waypalm/site/web/controllers/BaseController.java
package waypalm.site.web.controllers;

import com.iserv2.commons.lang.collections.IservCollections;
import org.springframework.beans.propertyeditors.StringTrimmerEditor;
import org.springframework.context.MessageSourceResolvable;
import org.springframework.security.authentication.InsufficientAuthenticationException;
import org.springframework.validation.Errors;
import org.springframework.validation.FieldError;
import org.springframework.validation.ObjectError;
import org.springframework.validation.SmartValidator;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.InitBinder;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.UrlBasedViewResolver;
import waypalm.common.services.FormatService;
import waypalm.common.web.model.ObjectErrors;
import waypalm.domain.entity.Profile;
import waypalm.domain.entity.World;
import waypalm.site.exceptions.ObjectNotFoundException;
import waypalm.site.services.ModelViewer;
import waypalm.site.services.UserAccess;

import javax.inject.Inject;
import java.io.Serializable;
import java.util.Locale;

public abstract class BaseController {
    @Inject
    protected UserAccess auth;
    @Inject
    protected FormatService format;
    @Inject
    protected ModelViewer viewer;
    @Inject
    protected SmartValidator validator;

    @InitBinder
    public void initBinder(WebDataBinder binder) {
        binder.registerCustomEditor(String.class, new StringTrimmerEditor(true));
    }


    public ModelAndView redirect(String url) {
        return view(UrlBasedViewResolver.REDIRECT_URL_PREFIX + url);
    }

    public ModelAndView forward(String url) {
        return view(UrlBasedViewResolver.FORWARD_URL_PREFIX + url);
    }

    public ModelAndView view(String viewName) {
        return new ModelAndView(viewName);
    }

    public ModelAndView json(Object val) {
        return view("jsonView").addObject("content", val);
    }

    protected ObjectErrors getValidationErrors(Locale locale, Errors formBinding) {
        if (formBinding == null || !formBinding.hasErrors()) {
            return null;
        }

        ObjectErrors formErrors = new ObjectErrors();
        for (FieldError formBindingFieldError : formBinding.getFieldErrors()) {
            ObjectErrors.FieldError fieldError = new ObjectErrors.FieldError();
            fieldError.setField(formBindingFieldError.getField());
            fieldError.setMessage(format.getMessage(locale, formBindingFieldError));
            formErrors.getFieldErrors().add(fieldError);
        }

        for (ObjectError formBindingFieldError : formBinding.getGlobalErrors()) {
            ObjectErrors.GlobalError fieldError = new ObjectErrors.GlobalError();
            fieldError.setMessage(format.getMessage(locale, formBindingFieldError));
            formErrors.getGlobalErrors().add(fieldError);
        }
        return formErrors;
    }

    protected String getErrorMessage(Locale locale, MessageSourceResolvable message) {
        if (message == null) {
            return null;
        }
        String code = IservCollections.firstOrNull(message.getCodes());
        return format.getMessage(locale, code == null ? "message.unknown" : code, message.getArguments());
    }

    public Profile requireProfile() {
        Profile profile = auth.loadCurrentProfile();
        if (profile == null) {
            throw new InsufficientAuthenticationException("require authenticated visitor");
        }
        return profile;
    }

    protected static <K extends Serializable> World requireEntity(World entity, K id) throws ObjectNotFoundException {
        if (entity == null)
            throw new ObjectNotFoundException(World.class, id);
        return entity;
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/controllers/BaseController.kt
package waypalm.site.web.controllers

import org.springframework.web.servlet.ModelAndView
import org.springframework.web.bind.annotation.InitBinder
import org.springframework.web.bind.WebDataBinder
import org.springframework.beans.propertyeditors.StringTrimmerEditor
import org.springframework.web.servlet.view.UrlBasedViewResolver
import waypalm.domain.entity.Profile
import org.springframework.security.authentication.InsufficientAuthenticationException
import waypalm.site.services.UserAccess
import javax.inject.Inject
import waypalm.common.web.model.ObjectErrors
import org.springframework.validation.Errors
import java.util.Locale
import org.springframework.context.MessageSourceResolvable
import waypalm.common.services.FormatService
import waypalm.common.util.firstSure
import waypalm.domain.entity.World
import java.io.Serializable
import waypalm.site.exceptions.ObjectNotFoundException
import waypalm.site.services.ModelViewer
import org.springframework.validation.SmartValidator

abstract class BaseController() {
    Inject
    protected var auth: UserAccess? = null
    Inject
    protected var format: FormatService? = null
    Inject
    protected var viewer: ModelViewer ? = null
    Inject
    protected var validator: SmartValidator? = null

    InitBinder
    public fun initBinder(binder: WebDataBinder) {
        binder.registerCustomEditor(javaClass<String>(), StringTrimmerEditor(true))
    }

    public fun view(name: String): ModelAndView {
        return ModelAndView(name);
    }

    public fun json(value: Unit): ModelAndView {
        return view("json").addObject("content", value)!!
    }

    public fun redirect(url: String): ModelAndView {
        return view(UrlBasedViewResolver.REDIRECT_URL_PREFIX + url);
    }

    public fun forward(url: String): ModelAndView {
        return view(UrlBasedViewResolver.FORWARD_URL_PREFIX + url);
    }

    public fun currentProfile(): Profile {
        var profile = auth!!.loadCurrentProfile();
        if (profile == null) {
            throw InsufficientAuthenticationException("require authenticated visitor");
        }
        return profile!!;
    }

    protected fun getValidationErrors(locale: Locale, formBinding: Errors?): ObjectErrors? {
        val formErrors = ObjectErrors();
        formBinding?.getFieldErrors()?.forEach {
            val fieldError = ObjectErrors.FieldError();
            fieldError.setField(it.getField());
            fieldError.setMessage(format!!.getMessage(locale, it));
            formErrors.getFieldErrors()!!.add(fieldError);
        }

        formBinding?.getGlobalErrors()?.forEach{
            var fieldError = ObjectErrors.GlobalError();
            fieldError.setMessage(format!!.getMessage(locale, it));
            formErrors.getGlobalErrors()!!.add(fieldError);
        }
        return formErrors;
    }

    protected fun getErrorMessage(locale: Locale, message: MessageSourceResolvable): String {
        var code = message.getCodes()!!.firstSure("message.unknown")
        return format!!.getMessage(locale, code, message.getArguments())!!;
    }

    protected fun requireEntity(entity: World?, id: Serializable): World {
        return if (entity == null) throw ObjectNotFoundException(javaClass<World>(), id) else entity;
    }
}